test/invariants/fuzz/BMath.t.sol

```diff
@@ -0,0 +1,93 @@
+
+  /**
+   * NOTE: These values were chosen to pass the fuzzing tests
+   * @dev Reducing BPOW_PRECISION may allow broader range of values increasing the gas cost
+   */
+  uint256 constant MAX_BALANCE = 1_000_000e18;
+  uint256 constant MIN_BALANCE = 1e18;
+  uint256 constant MAX_TOLERANCE = 1e12;
+
+  constructor() {
+    bMath = new BMath();
+
+    MIN_WEIGHT = bMath.MIN_WEIGHT();
+    MAX_WEIGHT = bMath.MAX_WEIGHT();
+    MIN_FEE = bMath.MIN_FEE();
+    MAX_FEE = bMath.MAX_FEE();
+  }
+
+  // calcOutGivenIn should be inverse of calcInGivenOut
+  function testCalcInGivenOut_InvCalcInGivenOut(
+    uint256 tokenBalanceIn,
+    uint256 tokenWeightIn,
+    uint256 tokenBalanceOut,
+    uint256 tokenWeightOut,
+    uint256 tokenAmountIn,
+    uint256 swapFee
+  ) public view {
+    tokenWeightIn = clamp(tokenWeightIn, MIN_WEIGHT, MAX_WEIGHT);
+    tokenWeightOut = clamp(tokenWeightOut, MIN_WEIGHT, MAX_WEIGHT);
+    tokenAmountIn = clamp(tokenAmountIn, MIN_BALANCE, MAX_BALANCE);
+    tokenBalanceOut = clamp(tokenBalanceOut, MIN_BALANCE, MAX_BALANCE);
+    tokenBalanceIn = clamp(tokenBalanceIn, MIN_BALANCE, MAX_BALANCE);
+    swapFee = clamp(swapFee, MIN_FEE, MAX_FEE);
+
+    uint256 calc_tokenAmountOut =
+      bMath.calcOutGivenIn(tokenBalanceIn, tokenWeightIn, tokenBalanceOut, tokenWeightOut, tokenAmountIn, swapFee);
+
+    uint256 calc_tokenAmountIn =
+      bMath.calcInGivenOut(tokenBalanceIn, tokenWeightIn, tokenBalanceOut, tokenWeightOut, calc_tokenAmountOut, swapFee);
+
+    assert(
+      calc_tokenAmountOut < MIN_BALANCE // return amount too small
+        || tokenAmountIn == calc_tokenAmountIn // exact match
+        || tokenAmountIn > calc_tokenAmountIn // within tolerance
+        ? (tokenAmountIn * 1e18 / calc_tokenAmountIn) - 1e18 <= MAX_TOLERANCE
+        : (calc_tokenAmountIn * 1e18 / tokenAmountIn) - 1e18 <= MAX_TOLERANCE
+    );
+  }
+
+  // calcInGivenOut should be inverse of calcOutGivenIn
+  function testCalcOutGivenIn_InvCalcOutGivenIn(
+    uint256 tokenBalanceIn,
+    uint256 tokenWeightIn,
+    uint256 tokenBalanceOut,
+    uint256 tokenWeightOut,
+    uint256 tokenAmountOut,
+    uint256 swapFee
+  ) public view {
+    tokenWeightIn = clamp(tokenWeightIn, MIN_WEIGHT, MAX_WEIGHT);
+    tokenWeightOut = clamp(tokenWeightOut, MIN_WEIGHT, MAX_WEIGHT);
+    tokenAmountOut = clamp(tokenAmountOut, MIN_BALANCE, MAX_BALANCE);
+    tokenBalanceOut = clamp(tokenBalanceOut, MIN_BALANCE, MAX_BALANCE);
+    tokenBalanceIn = clamp(tokenBalanceIn, MIN_BALANCE, MAX_BALANCE);
+    swapFee = clamp(swapFee, MIN_FEE, MAX_FEE);
+
+    uint256 calc_tokenAmountIn =
+      bMath.calcInGivenOut(tokenBalanceOut, tokenWeightOut, tokenBalanceIn, tokenWeightIn, tokenAmountOut, swapFee);
+
+    uint256 calc_tokenAmountOut =
+      bMath.calcOutGivenIn(tokenBalanceOut, tokenWeightOut, tokenBalanceIn, tokenWeightIn, calc_tokenAmountIn, swapFee);
+
+    assert(
+      calc_tokenAmountIn < MIN_BALANCE // return amount too small
```

Comment:
I assume the purpose of this check is to skip the assertion if the output is too small which drops us into some edge case where the error is greater, but while testing it I don't seem to find cases where it breaks (perhaps I didn't stumble upon that case *yet* though)

Also I think this discards a whole lot of cases which are valid, I'd remove it if no counter-examples are found, or otherwise would try to use a lower value and try to find out why it breaks

Reply:
if we're making a minimum amount to be traded, this assert is making sure that if the output is below that amount we should also discard the case